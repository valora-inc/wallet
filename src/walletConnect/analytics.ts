import { SessionTypes } from '@walletconnect/types'
import { Web3WalletTypes } from '@walletconnect/web3wallet'
import { getDappRequestOrigin } from 'src/app/utils'
import { ActiveDapp } from 'src/dapps/types'

const isSessionProposalType = (
  session: Web3WalletTypes.EventArguments['session_proposal'] | SessionTypes.Struct
): session is Web3WalletTypes.EventArguments['session_proposal'] => {
  return 'params' in session
}

export function getDefaultSessionTrackedProperties(
  session: Web3WalletTypes.EventArguments['session_proposal'] | SessionTypes.Struct,
  activeDapp: ActiveDapp | null
) {
  const peer = isSessionProposalType(session) ? session.params.proposer : session.peer
  const { name: dappName, url: dappUrl, description: dappDescription, icons } = peer.metadata

  const relayProtocol = isSessionProposalType(session)
    ? session.params.relays[0]?.protocol
    : session.relay.protocol

  const requiredNamespaces = isSessionProposalType(session)
    ? session.params.requiredNamespaces
    : session.requiredNamespaces

  return {
    version: 2 as const,
    dappRequestOrigin: getDappRequestOrigin(activeDapp),
    dappName,
    dappUrl,
    dappDescription,
    dappIcon: icons[0],
    relayProtocol,
    ...Object.keys(requiredNamespaces).reduce((acc, key) => {
      const { chains, events, methods } = requiredNamespaces[key]
      return {
        ...acc,
        [`${key}Events`]: events,
        [`${key}Chains`]: chains,
        [`${key}Methods`]: methods,
      }
    }, {}),
  }
}

export function getDefaultRequestTrackedProperties(
  request: Web3WalletTypes.EventArguments['session_request']
) {
  const { id, params } = request

  return {
    requestChainId: params.chainId,
    requestId: id,
    requestJsonrpc: '2.0',
    requestMethod: params.request.method,
  }
}
