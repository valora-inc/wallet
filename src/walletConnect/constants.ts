import { TFunction } from 'i18next'

export enum SupportedActions {
  eth_accounts = 'eth_accounts',
  eth_signTransaction = 'eth_signTransaction',
  eth_sendTransaction = 'eth_sendTransaction',
  eth_signTypedData = 'eth_signTypedData',
  eth_signTypedData_v4 = 'eth_signTypedData_v4',
  eth_sign = 'eth_sign',
  personal_sign = 'personal_sign',
  personal_decrypt = 'personal_decrypt',
  computeSharedSecret = 'personal_computeSharedSecret',
}

export function isSupportedAction(action: string) {
  return Object.values(SupportedActions).includes(action as SupportedActions)
}

export function getTranslationDescriptionFromAction(t: TFunction, action: SupportedActions) {
  const actionDescriptionTranslations: { [action in SupportedActions]: string } = {
    [SupportedActions.eth_accounts]: t('description.accounts'),
    [SupportedActions.eth_signTransaction]: t('description.signTransaction'),
    [SupportedActions.eth_sendTransaction]: t('description.sendTransaction'),
    [SupportedActions.eth_signTypedData]: t('description.sign'),
    [SupportedActions.eth_signTypedData_v4]: t('description.sign'),
    [SupportedActions.eth_sign]: t('description.sign'),
    [SupportedActions.personal_sign]: t('description.sign'),
    [SupportedActions.personal_decrypt]: t('description.decrypt'),
    [SupportedActions.computeSharedSecret]: t('description.computeSharedSecret'),
  }

  const translationId = actionDescriptionTranslations[action]
  if (!translationId) {
    return ''
  }

  return translationId
}

export function getTranslationFromAction(t: TFunction, action: SupportedActions) {
  const actionTranslations: { [x in SupportedActions]: string } = {
    [SupportedActions.eth_accounts]: t('action.accounts'),
    [SupportedActions.eth_signTransaction]: t('action.signTransaction'),
    [SupportedActions.eth_sendTransaction]: t('action.sendTransaction'),
    [SupportedActions.eth_signTypedData]: t('action.sign'),
    [SupportedActions.eth_signTypedData_v4]: t('action.sign'),
    [SupportedActions.eth_sign]: t('action.sign'),
    [SupportedActions.personal_sign]: t('action.sign'),
    [SupportedActions.personal_decrypt]: t('action.decrypt'),
    [SupportedActions.computeSharedSecret]: t('action.computeSharedSecret'),
  }

  const translationId = actionTranslations[action]
  if (!translationId) {
    return ''
  }

  return translationId
}
