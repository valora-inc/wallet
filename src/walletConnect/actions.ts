import { JsonRpcTypes, SessionTypes } from '@walletconnect/types'
import { Web3WalletTypes } from '@walletconnect/web3wallet'
import { WalletConnectPairingOrigin } from 'src/analytics/types'
import { TransactionRequest } from 'src/viem/prepareTransactions'

export enum Actions {
  /**
   * Actions coming as a result of user action
   */
  INITIALISE_CLIENT = 'WALLETCONNECT/INITIALISE_CLIENT',
  INITIALISE_PAIRING = 'WALLETCONNECT/INITIALISE_PAIRING',

  ACCEPT_SESSION = 'WALLETCONNECT/ACCEPT_SESSION',
  DENY_SESSION = 'WALLETCONNECT/DENY_SESSION',
  CLOSE_SESSION = 'WALLETCONNECT/CLOSE_SESSION',
  CLOSE_PENDING_SESSION = 'WALLETCONNECT/CLOSE_PENDING_SESSION',

  SHOW_REQUEST_DETAILS = 'WALLETCONNECT/SHOW_REQUEST_DETAILS',
  ACCEPT_REQUEST = 'WALLETCONNECT/ACCEPT_REQUEST',
  DENY_REQUEST = 'WALLETCONNECT/DENY_REQUEST',
  REQUEST_FULFILLED = 'WALLETCONNECT/REQUEST_FULFILLED',

  CLIENT_INITIALISED = 'WALLETCONNECT/CLIENT_INITIALISED',
  CLIENT_DESTROYED = 'WALLETCONNECT/CLIENT_DESTROYED',

  REMOVE_EXPIRED_SESSIONS = 'WALLETCONNECT/REMOVE_EXPIRED_SESSIONS',

  /**
   * Actions coming from the WalletConnect client
   */
  SESSION_PROPOSAL = 'WALLETCONNECT/SESSION_PROPOSAL',
  SESSION_CREATED = 'WALLETCONNECT/SESSION_CREATED',
  SESSION_DELETED = 'WALLETCONNECT/SESSION_DELETED',
  SESSION_PAYLOAD = 'WALLETCONNECT/SESSION_PAYLOAD',
}

export interface InitialiseClient {
  type: Actions.INITIALISE_CLIENT
}
export interface ClientInitialised {
  type: Actions.CLIENT_INITIALISED
}
export interface ClientDestroyed {
  type: Actions.CLIENT_DESTROYED
}

/**
 * Session objects
 */
export interface AcceptSession {
  type: Actions.ACCEPT_SESSION
  session: Web3WalletTypes.EventArguments['session_proposal']
  approvedNamespaces: SessionTypes.Namespaces
}

export interface DenySession {
  type: Actions.DENY_SESSION
  session: Web3WalletTypes.EventArguments['session_proposal']
  reason: JsonRpcTypes.Error
}
export interface CloseSession {
  type: Actions.CLOSE_SESSION
  session: SessionTypes.Struct
}
export interface ShowRequestDetails {
  type: Actions.SHOW_REQUEST_DETAILS
  request: Web3WalletTypes.EventArguments['session_request']
  infoString: string
}
export interface AcceptRequest {
  type: Actions.ACCEPT_REQUEST
  request: Web3WalletTypes.EventArguments['session_request']
  preparedTransactions?: TransactionRequest[]
}
export interface DenyRequest {
  type: Actions.DENY_REQUEST
  request: Web3WalletTypes.EventArguments['session_request']
  reason: JsonRpcTypes.Error
}

export interface RemoveExpiredSessions {
  type: Actions.REMOVE_EXPIRED_SESSIONS
  dateInSeconds: number
}

export interface InitialisePairing {
  type: Actions.INITIALISE_PAIRING
  uri: string
  origin: WalletConnectPairingOrigin
}

export interface SessionProposal {
  type: Actions.SESSION_PROPOSAL
  session: Web3WalletTypes.EventArguments['session_proposal']
}
export interface SessionCreated {
  type: Actions.SESSION_CREATED
  session: SessionTypes.Struct
}

export interface SessionDeleted {
  type: Actions.SESSION_DELETED
  session: Web3WalletTypes.EventArguments['session_delete']
}
export interface SessionPayload {
  type: Actions.SESSION_PAYLOAD
  request: Web3WalletTypes.EventArguments['session_request']
}

export type WalletConnectActions =
  | SessionProposal
  | SessionCreated
  | SessionDeleted
  | SessionPayload

export type UserActions =
  | InitialiseClient
  | InitialisePairing
  | ClientInitialised
  | ClientDestroyed
  | AcceptSession
  | DenySession
  | CloseSession
  | ShowRequestDetails
  | AcceptRequest
  | DenyRequest
  | RemoveExpiredSessions

export const initialiseClient = (): InitialiseClient => ({
  type: Actions.INITIALISE_CLIENT,
})

export const initialisePairing = (
  uri: string,
  origin: WalletConnectPairingOrigin
): InitialisePairing => ({
  type: Actions.INITIALISE_PAIRING,
  uri,
  origin,
})

export const acceptSession = (
  session: Web3WalletTypes.EventArguments['session_proposal'],
  approvedNamespaces: SessionTypes.Namespaces
): AcceptSession => ({
  type: Actions.ACCEPT_SESSION,
  session,
  approvedNamespaces,
})

export const denySession = (
  session: Web3WalletTypes.EventArguments['session_proposal'],
  reason: JsonRpcTypes.Error
): DenySession => ({
  type: Actions.DENY_SESSION,
  session,
  reason,
})

export const closeSession = (session: SessionTypes.Struct): CloseSession => ({
  type: Actions.CLOSE_SESSION,
  session,
})

export const showRequestDetails = (
  request: Web3WalletTypes.EventArguments['session_request'],
  infoString: string
): ShowRequestDetails => ({
  type: Actions.SHOW_REQUEST_DETAILS,
  request,
  infoString,
})

export const acceptRequest = (
  request: Web3WalletTypes.EventArguments['session_request'],
  preparedTransactions?: TransactionRequest[]
): AcceptRequest => ({
  type: Actions.ACCEPT_REQUEST,
  request,
  preparedTransactions,
})

export const denyRequest = (
  request: Web3WalletTypes.EventArguments['session_request'],
  reason: JsonRpcTypes.Error
): DenyRequest => ({
  type: Actions.DENY_REQUEST,
  request,
  reason,
})

export const removeExpiredSessions = (dateInSeconds: number): RemoveExpiredSessions => ({
  type: Actions.REMOVE_EXPIRED_SESSIONS,
  dateInSeconds,
})

export const clientInitialised = (): ClientInitialised => ({
  type: Actions.CLIENT_INITIALISED,
})

export const clientDestroyed = (): ClientDestroyed => ({
  type: Actions.CLIENT_DESTROYED,
})

export const sessionProposal = (
  session: Web3WalletTypes.EventArguments['session_proposal']
): SessionProposal => ({
  type: Actions.SESSION_PROPOSAL,
  session,
})

export const sessionCreated = (session: SessionTypes.Struct): SessionCreated => ({
  type: Actions.SESSION_CREATED,
  session,
})

export const sessionDeleted = (
  session: Web3WalletTypes.EventArguments['session_delete']
): SessionDeleted => ({
  type: Actions.SESSION_DELETED,
  session,
})

export const sessionPayload = (
  request: Web3WalletTypes.EventArguments['session_request']
): SessionPayload => ({
  type: Actions.SESSION_PAYLOAD,
  request,
})
