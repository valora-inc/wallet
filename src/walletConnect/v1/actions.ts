import { WalletConnectPairingOrigin } from 'src/analytics/types'
import {
  WalletConnectPayloadRequest,
  WalletConnectSession,
  WalletConnectSessionRequest,
} from 'src/walletConnect/types'

export enum Actions {
  /**
   * Actions coming as a result of user action
   */
  INITIALISE_CONNECTION_V1 = 'WALLETCONNECT/INITIALISE_CONNECTION_V1',

  ACCEPT_SESSION_V1 = 'WALLETCONNECT/ACCEPT_SESSION_V1',
  DENY_SESSION_V1 = 'WALLETCONNECT/DENY_SESSION_V1',
  CLOSE_SESSION_V1 = 'WALLETCONNECT/CLOSE_SESSION_V1',

  SHOW_REQUEST_DETAILS_V1 = 'WALLETCONNECT/SHOW_REQUEST_DETAILS_V1',
  ACCEPT_REQUEST_V1 = 'WALLETCONNECT/ACCEPT_REQUEST_V1',
  DENY_REQUEST_V1 = 'WALLETCONNECT/DENY_REQUEST_V1',
  REQUEST_HANDLED_V1 = 'WALLETCONNECT/REQUEST_HANDLED_V1',

  STORE_SESSION_V1 = 'WALLETCONNECT/STORE_SESSION_V1',

  /**
   * Actions coming from the WalletConnect client
   */
  SESSION_V1 = 'WALLETCONNECT/SESSION_V1',
  SESSION_DELETED_V1 = 'WALLETCONNECT/SESSION_DELETED_V1',
  PAYLOAD_V1 = 'WALLETCONNECT/REQUEST_V1',
}

export interface InitialiseConnection {
  type: Actions.INITIALISE_CONNECTION_V1
  uri: string
  origin: WalletConnectPairingOrigin
}

/**
 * Session objects
 */
export interface AcceptSession {
  type: Actions.ACCEPT_SESSION_V1
  peerId: string
  session: WalletConnectSessionRequest
}
export interface DenySession {
  type: Actions.DENY_SESSION_V1
  peerId: string
  session: WalletConnectSessionRequest
}
export interface CloseSession {
  type: Actions.CLOSE_SESSION_V1
  session: WalletConnectSession
}
export interface StoreSession {
  type: Actions.STORE_SESSION_V1
  session: WalletConnectSession
}
export interface ShowRequestDetails {
  type: Actions.SHOW_REQUEST_DETAILS_V1
  request: WalletConnectPayloadRequest
  peerId: string
  infoString: string
}
export interface AcceptRequest {
  type: Actions.ACCEPT_REQUEST_V1
  request: WalletConnectPayloadRequest
  peerId: string
}
export interface DenyRequest {
  type: Actions.DENY_REQUEST_V1
  request: WalletConnectPayloadRequest
  peerId: string
  reason: string
}

export interface SessionRequest {
  type: Actions.SESSION_V1
  peerId: string
  session: WalletConnectSessionRequest
}
export interface SessionDeleted {
  type: Actions.SESSION_DELETED_V1
  peerId: string
}
export interface PayloadRequest {
  type: Actions.PAYLOAD_V1
  peerId: string
  request: WalletConnectPayloadRequest
}

export type WalletConnectActions = SessionRequest | PayloadRequest

export type UserActions =
  | AcceptSession
  | DenySession
  | SessionDeleted
  | StoreSession
  | CloseSession
  | ShowRequestDetails
  | AcceptRequest
  | DenyRequest

export const initialiseConnection = (
  uri: string,
  origin: WalletConnectPairingOrigin
): InitialiseConnection => ({
  type: Actions.INITIALISE_CONNECTION_V1,
  uri,
  origin,
})

export const acceptSession = (session: WalletConnectSessionRequest): AcceptSession => ({
  type: Actions.ACCEPT_SESSION_V1,
  peerId: session.params[0].peerId,
  session,
})
export const denySession = (session: WalletConnectSessionRequest): DenySession => ({
  type: Actions.DENY_SESSION_V1,
  peerId: session.params[0].peerId,
  session,
})
export const closeSession = (session: WalletConnectSession) => ({
  type: Actions.CLOSE_SESSION_V1,
  session,
})
export const storeSession = (session: WalletConnectSession) => ({
  type: Actions.STORE_SESSION_V1,
  session,
})

export const showRequestDetails = (
  peerId: string,
  request: WalletConnectPayloadRequest,
  infoString: string
): ShowRequestDetails => ({
  type: Actions.SHOW_REQUEST_DETAILS_V1,
  request,
  peerId,
  infoString,
})
export const acceptRequest = (
  peerId: string,
  request: WalletConnectPayloadRequest
): AcceptRequest => ({
  type: Actions.ACCEPT_REQUEST_V1,
  request,
  peerId,
})
export const denyRequest = (
  peerId: string,
  request: WalletConnectPayloadRequest,
  reason: string
): DenyRequest => ({
  type: Actions.DENY_REQUEST_V1,
  request,
  peerId,
  reason,
})

export const sessionRequest = (
  peerId: string,
  session: WalletConnectSessionRequest
): SessionRequest => ({
  type: Actions.SESSION_V1,
  session,
  peerId,
})
export const sessionDeleted = (peerId: string): SessionDeleted => ({
  type: Actions.SESSION_DELETED_V1,
  peerId,
})
export const payloadRequest = (
  peerId: string,
  request: WalletConnectPayloadRequest
): PayloadRequest => ({
  type: Actions.PAYLOAD_V1,
  peerId,
  request,
})
