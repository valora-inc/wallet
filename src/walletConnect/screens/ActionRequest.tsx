import { trimLeading0x } from '@celo/utils/lib/address'
import { StackScreenProps } from '@react-navigation/stack'
import React, { useEffect } from 'react'
import { useTranslation } from 'react-i18next'
import { Image, ScrollView, StyleSheet, Text, View } from 'react-native'
import { TouchableOpacity } from 'react-native-gesture-handler'
import { SafeAreaView } from 'react-native-safe-area-context'
import { useDispatch } from 'react-redux'
import Button, { BtnSizes, BtnTypes } from 'src/components/Button'
import { headerWithCloseButton } from 'src/navigator/Headers'
import { Screens } from 'src/navigator/Screens'
import { StackParamList } from 'src/navigator/types'
import colors from 'src/styles/colors'
import fontStyles from 'src/styles/fonts'
import useStateWithCallback from 'src/utils/useStateWithCallback'
import { getTranslationFromAction, SupportedActions } from 'src/walletConnect/constants'
import {
  acceptRequest as acceptRequestV1,
  denyRequest as denyRequestV1,
  showRequestDetails as showRequestDetailsV1,
} from 'src/walletConnect/v1/actions'

type Props = StackScreenProps<StackParamList, Screens.WalletConnectActionRequest>

function showRequestDetails(params: Props['route']['params'], infoString: string) {
  return showRequestDetailsV1(params.peerId, params.action, infoString)
}

function acceptRequest(params: Props['route']['params']) {
  return acceptRequestV1(params.peerId, params.action)
}

function denyRequest(params: Props['route']['params']) {
  return denyRequestV1(params.peerId, params.action, 'User denied')
}

function getRequestInfo(params: Props['route']['params']) {
  return {
    url: params.dappUrl,
    name: params.dappName,
    icon: params.dappIcon,
    method: params.action.method,
    params: params.action.params,
  }
}
function ActionRequest({ navigation, route: { params: routeParams } }: Props) {
  const { t } = useTranslation()
  const [isAccepting, setIsAccepting] = useStateWithCallback(false)
  const [isDenying, setIsDenying] = useStateWithCallback(false)
  const dispatch = useDispatch()

  const onAccept = () => {
    // Dispatch after state has been changed to avoid triggering the 'beforeRemove' action while processing
    setIsAccepting(true, () => dispatch(acceptRequest(routeParams)))
  }

  const onDeny = () => {
    // Dispatch after state has been changed to avoid triggering the 'beforeRemove' action while processing
    setIsDenying(true, () => dispatch(denyRequest(routeParams)))
  }

  const isLoading = isAccepting || isDenying

  useEffect(
    () =>
      navigation.addListener('beforeRemove', (e) => {
        if (isLoading) {
          return
        }
        e.preventDefault()
        onDeny()
      }),
    [navigation, routeParams, isLoading]
  )

  const { url, name, icon, method, params } = getRequestInfo(routeParams)
  const moreInfoString =
    method === SupportedActions.eth_signTransaction ||
    method === SupportedActions.eth_sendTransaction
      ? JSON.stringify(params)
      : method === SupportedActions.eth_signTypedData ||
        method === SupportedActions.eth_signTypedData_v4
      ? JSON.stringify(params[1])
      : method === SupportedActions.personal_decrypt
      ? Buffer.from(params[1]).toString('hex')
      : method === SupportedActions.personal_sign
      ? Buffer.from(trimLeading0x(params[0]), 'hex').toString() ||
        params[0] ||
        t('action.emptyMessage')
      : null

  const onMoreInfo = () => {
    if (!moreInfoString) {
      return
    }

    dispatch(showRequestDetails(routeParams, moreInfoString))
  }

  const uri = icon ?? `${url}/favicon.ico`

  return (
    <SafeAreaView style={styles.container}>
      <ScrollView contentContainerStyle={styles.scrollContainer}>
        <View style={styles.center}>
          <Image style={styles.logo} source={{ uri }} />
        </View>
        <Text style={styles.header}>{t('connectToWallet', { dappName: name })}</Text>
        <Text style={styles.share}>{t('action.asking')}:</Text>

        <View style={styles.sectionDivider}>
          <Text style={styles.sectionHeaderText}>{t('action.operation')}</Text>
          <Text style={styles.bodyText}>
            {getTranslationFromAction(t, method as SupportedActions)}
          </Text>

          {moreInfoString && (
            <>
              <Text style={styles.sectionHeaderText}>{t('action.data')}</Text>
              <TouchableOpacity onPress={onMoreInfo}>
                <Text style={[styles.bodyText, styles.underLine]}>{t('action.details')}</Text>
              </TouchableOpacity>
            </>
          )}
        </View>

        <View style={styles.buttonContainer} pointerEvents={isLoading ? 'none' : undefined}>
          <Button
            style={styles.button}
            type={BtnTypes.SECONDARY}
            size={BtnSizes.MEDIUM}
            text={t('cancel')}
            showLoading={isDenying}
            onPress={onDeny}
            testID="WalletConnectActionCancel"
          />
          <Button
            style={styles.button}
            type={BtnTypes.PRIMARY}
            size={BtnSizes.MEDIUM}
            text={t('allow')}
            showLoading={isAccepting}
            onPress={onAccept}
            testID="WalletConnectActionAllow"
          />
        </View>
      </ScrollView>
    </SafeAreaView>
  )
}

ActionRequest.navigationOptions = headerWithCloseButton

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  center: {
    alignItems: 'center',
  },
  header: {
    ...fontStyles.h1,
    textAlign: 'center',
    paddingVertical: 16,
  },
  logo: {
    height: 80,
    width: 80,
  },
  share: {
    ...fontStyles.regular,
    color: colors.gray4,
    textAlign: 'center',
  },
  sectionDivider: {
    alignItems: 'center',
    width: 200,
  },
  sectionHeaderText: {
    ...fontStyles.label,
    marginTop: 16,
    marginBottom: 4,
  },
  button: {
    marginTop: 24,
  },
  scrollContainer: {
    flexGrow: 1,
    alignItems: 'center',
    justifyContent: 'center',
    marginHorizontal: '15%',
  },
  buttonContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-around',
    width: '100%',
  },
  bodyText: {
    ...fontStyles.regular,
    color: colors.gray4,
    textAlign: 'center',
  },
  underLine: {
    textDecorationLine: 'underline',
  },
})

export default ActionRequest
