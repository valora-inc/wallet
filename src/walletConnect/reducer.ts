import { combineReducers } from 'redux'
import { reducer as reducerV1, State as stateV1 } from 'src/walletConnect/v1/reducer'
import { reducer as reducerV2, State as stateV2 } from 'src/walletConnect/v2/reducer'

export interface State {
  v1: stateV1
  v2: stateV2
}

export const reducer = combineReducers({ v1: reducerV1, v2: reducerV2 })
