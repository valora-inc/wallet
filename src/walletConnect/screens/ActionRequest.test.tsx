import { fireEvent, render } from '@testing-library/react-native'
import * as React from 'react'
import 'react-native'
import { Provider } from 'react-redux'
import { Screens } from 'src/navigator/Screens'
import { SupportedActions } from 'src/walletConnect/constants'
import ActionRequest from 'src/walletConnect/screens/ActionRequest'
import { Actions } from 'src/walletConnect/v1/actions'
import { createMockStore, getMockStackScreenProps } from 'test/utils'

describe('ActionRequest', () => {
  const store = createMockStore({})

  afterEach(() => {
    store.clearActions()
  })

  describe('personal_sign', () => {
    const action = {
      id: 1,
      jsonrpc: '',
      method: SupportedActions.personal_sign,
      params: [
        '0x4d65737361676520746f207369676e', // hex of 'Message to sign'
        '0xe17becad62a0a1225473bb52e620ae29728b55a0',
      ],
    }
    it('renders correctly', () => {
      const { toJSON } = render(
        <Provider store={store}>
          <ActionRequest
            {...getMockStackScreenProps(Screens.WalletConnectActionRequest, {
              dappName: 'foo',
              dappIcon: 'foo',
              dappUrl: 'foo',
              action,
              version: 1,
              peerId: 'peerId',
            })}
          />
        </Provider>
      )
      expect(toJSON()).toMatchSnapshot()
    })

    it('dispatches request details with correct string on clicking details', async () => {
      const { getByText } = render(
        <Provider store={store}>
          <ActionRequest
            {...getMockStackScreenProps(Screens.WalletConnectActionRequest, {
              dappName: 'foo',
              dappIcon: 'foo',
              dappUrl: 'foo',
              action,
              version: 1,
              peerId: 'peerId',
            })}
          />
        </Provider>
      )
      await fireEvent.press(getByText('action.details'))
      expect(store.getActions()).toEqual([
        {
          type: Actions.SHOW_REQUEST_DETAILS_V1,
          request: action,
          peerId: 'peerId',
          infoString: 'Message to sign',
        },
      ])
    })

    it('dispatches request details with raw string if message cannot be decoded', async () => {
      action.params[0] = 'invalid hex'
      const { getByText } = render(
        <Provider store={store}>
          <ActionRequest
            {...getMockStackScreenProps(Screens.WalletConnectActionRequest, {
              dappName: 'foo',
              dappIcon: 'foo',
              dappUrl: 'foo',
              action,
              version: 1,
              peerId: 'peerId',
            })}
          />
        </Provider>
      )
      await fireEvent.press(getByText('action.details'))
      expect(store.getActions()).toEqual([
        {
          type: Actions.SHOW_REQUEST_DETAILS_V1,
          request: action,
          peerId: 'peerId',
          infoString: 'invalid hex',
        },
      ])
    })

    it('dispatches request details with empty message', async () => {
      action.params[0] = ''
      const { getByText } = render(
        <Provider store={store}>
          <ActionRequest
            {...getMockStackScreenProps(Screens.WalletConnectActionRequest, {
              dappName: 'foo',
              dappIcon: 'foo',
              dappUrl: 'foo',
              action,
              version: 1,
              peerId: 'peerId',
            })}
          />
        </Provider>
      )
      await fireEvent.press(getByText('action.details'))
      expect(store.getActions()).toEqual([
        {
          type: Actions.SHOW_REQUEST_DETAILS_V1,
          request: action,
          peerId: 'peerId',
          infoString: 'action.emptyMessage',
        },
      ])
    })
  })
})
