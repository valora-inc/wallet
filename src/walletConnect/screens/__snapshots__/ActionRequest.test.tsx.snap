// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionRequest personal_sign renders correctly 1`] = `
<RNCSafeAreaView
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "alignItems": "center",
        "flexGrow": 1,
        "justifyContent": "center",
        "marginHorizontal": "15%",
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "foo",
            }
          }
          style={
            Object {
              "height": 80,
              "width": 80,
            }
          }
        />
      </View>
      <Text
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Jost-Book",
            "fontSize": 26,
            "lineHeight": 32,
            "paddingVertical": 16,
            "textAlign": "center",
          }
        }
      >
        connectToWallet, {"dappName":"foo"}
      </Text>
      <Text
        style={
          Object {
            "color": "#9CA4A9",
            "fontFamily": "Inter-Regular",
            "fontSize": 16,
            "lineHeight": 22,
            "textAlign": "center",
          }
        }
      >
        action.asking
        :
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "width": 200,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-SemiBold",
              "fontSize": 13,
              "lineHeight": 16,
              "marginBottom": 4,
              "marginTop": 16,
            }
          }
        >
          action.operation
        </Text>
        <Text
          style={
            Object {
              "color": "#9CA4A9",
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "lineHeight": 22,
              "textAlign": "center",
            }
          }
        >
          action.sign
        </Text>
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-SemiBold",
              "fontSize": 13,
              "lineHeight": 16,
              "marginBottom": 4,
              "marginTop": 16,
            }
          }
        >
          action.data
        </Text>
        <RNGestureHandlerButton
          collapsable={false}
          onGestureEvent={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onHandlerStateChange={[Function]}
          rippleColor={0}
        >
          <View
            accessible={true}
            style={
              Array [
                Object {},
                Object {
                  "opacity": 1,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#9CA4A9",
                    "fontFamily": "Inter-Regular",
                    "fontSize": 16,
                    "lineHeight": 22,
                    "textAlign": "center",
                  },
                  Object {
                    "textDecorationLine": "underline",
                  },
                ]
              }
            >
              action.details
            </Text>
          </View>
        </RNGestureHandlerButton>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-around",
            "width": "100%",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
              },
              Object {
                "marginTop": 24,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 100,
                },
              ]
            }
          >
            <View
              accessible={true}
              focusable={true}
              nativeBackgroundAndroid={
                Object {
                  "attribute": "selectableItemBackground",
                  "rippleRadius": undefined,
                  "type": "ThemeAttrAndroid",
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#F1F0EB",
                  "height": 48,
                  "justifyContent": "center",
                  "minWidth": 120,
                  "opacity": undefined,
                  "paddingHorizontal": 24,
                  "paddingVertical": 5,
                }
              }
              testID="WalletConnectActionCancel"
            >
              <Text
                maxFontSizeMultiplier={1}
                style={
                  Object {
                    "color": "#2E3338",
                    "fontFamily": "Inter-SemiBold",
                    "fontSize": 16,
                    "lineHeight": 22,
                  }
                }
              >
                cancel
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "flexDirection": "row",
              },
              Object {
                "marginTop": 24,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 100,
                },
              ]
            }
          >
            <View
              accessible={true}
              focusable={true}
              nativeBackgroundAndroid={
                Object {
                  "attribute": "selectableItemBackground",
                  "rippleRadius": undefined,
                  "type": "ThemeAttrAndroid",
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#1AB775",
                  "height": 48,
                  "justifyContent": "center",
                  "minWidth": 120,
                  "opacity": undefined,
                  "paddingHorizontal": 24,
                  "paddingVertical": 5,
                }
              }
              testID="WalletConnectActionAllow"
            >
              <Text
                maxFontSizeMultiplier={1}
                style={
                  Object {
                    "color": "#FFFFFF",
                    "fontFamily": "Inter-SemiBold",
                    "fontSize": 16,
                    "lineHeight": 22,
                  }
                }
              >
                allow
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;
